LibHR/Geometry/communications.c

```diff
@@ -1,1103 +1,150 @@
-  complete_sc_sendrecv(sf);
-
-  /* fill send buffers */
-  sync_suNg_scalar_field(sf);
-#ifdef MPI_TIMING
-  error(sf_control > 0, 1, "start_sc_sendrecv " __FILE__, "Multiple send without receive");
-  gettimeofday(&sfstart, 0);
-  sf_control = 1;
-#endif
-
-  int shift = gd->master_shift;
-#ifdef WITH_NEW_GEOMETRY
-  shift = 0;
-#endif
-
-  for (i = 0; i < (gd->nbuffers_spinor); ++i)
-  {
-    /* send ith buffer */
-    mpiret = MPI_Isend((double *)((sf->sendbuf_ptr) + (gd->sbuf_start[i]) - shift), /* buffer */
-                       (gd->sbuf_len[i]) * (sizeof(suNg_vector) / sizeof(double)),       /* lenght in units of doubles */
-                       MPI_DOUBLE,                                                       /* basic datatype */
-                       gd->sbuf_to_proc[i],                                              /* cid of destination */
-                       i,                                                                /* tag of communication */
-                       cart_comm,                                                        /* use the cartesian communicator */
-                       &(sf->comm_req[2 * i])                                            /* handle to communication request */
-    );
-#ifndef NDEBUG
-    if (mpiret != MPI_SUCCESS)
-    {
-      char mesg[MPI_MAX_ERROR_STRING];
-      int mesglen;
-      MPI_Error_string(mpiret, mesg, &mesglen);
-      lprintf("MPI", 0, "ERROR: %s\n", mesg);
-      error(1, 1, "start_sc_sendrecv " __FILE__, "Cannot start send buffer");
-    }
-#endif
-
-    /* receive ith buffer */
-    mpiret = MPI_Irecv((double *)((sf->ptr) + (gd->rbuf_start[i]) - (gd->master_shift)), /* buffer */
-                       (gd->rbuf_len[i]) * (sizeof(suNg_vector) / sizeof(double)),       /* lenght in units of doubles */
-                       MPI_DOUBLE,                                                       /* basic datatype */
-                       gd->rbuf_from_proc[i],                                            /* cid of origin */
-                       i,                                                                /* tag of communication */
-                       cart_comm,                                                        /* use the cartesian communicator */
-                       &(sf->comm_req[2 * i + 1])                                        /* handle to communication request */
-    );
-#ifndef NDEBUG
-    if (mpiret != MPI_SUCCESS)
-    {
-      char mesg[MPI_MAX_ERROR_STRING];
-      int mesglen;
-      MPI_Error_string(mpiret, mesg, &mesglen);
-      lprintf("MPI", 0, "ERROR: %s\n", mesg);
-      error(1, 1, "start_sc_sendrecv " __FILE__, "Cannot start receive buffer");
-    }
-#endif
-  }
-
-#endif /* WITH_MPI */
-}
-
+#include "io.h"
+#include "random.h"
+
+// TODO: put gpu as last suffix
+// TODO: fill buffers needs gpu suffix
+// TODO: MPI error management
+    
+#define random_double ranlxd
+#define random_float ranlxs
```

Comment:
what are these for?

Reply:
Just so that in the template below I can use the real type given as _REAL to construct the right function name. These are not central to the communications but for the fill buffer functions, which, at the moment, I only use for testing. 

This is line 171 in the corresponding template. They are, however, supposed to be only local, so it would probably be best if I moved this into the template and undefine at the end of the file.

Reply:
ok I see. I think it is fine for now, but we should really use _Generic for this.